### ISSUE TITLE
Still receiving `instantiate`: undefined method `id` for nil:nilClass on 0.7.0

### ISSUE THREAD

**Issue by asciifaceman** [chef-partners/knife-vcair#33]
I went through a few other issues referencing this and have included a few notes from it. I am using VPC on-demand and have configured appropriately. 

```
( ccorbett@cherenkov:~/development/cookbooks/sce-vcloud (feature) )$ bash test.sh 
INFO: Using configuration from /home/ccorbett/.chef/knife.rb
INFO: Started chef-zero at chefzero://localhost:8889 with repository at /home/ccorbett/development
  One version per cookbook

DEBUG: vcair_username OBFUSCATE
DEBUG: vcair_org OBFUSCATE
DEBUG: vcair_api_host OBFUSCATE
DEBUG: vcair_api_path /api/compute/api
DEBUG: vcair_api_version 
DEBUG: vcair_show_progress 
DEBUG: Using fog version: 1.34.0
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/knife-vcair-0.7.0/lib/chef/knife/vcair_server_create.rb:148:in `instantiate': undefined method `id' for nil:NilClass (NoMethodError)
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/knife-vcair-0.7.0/lib/chef/knife/vcair_server_create.rb:36:in `execute_command'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/knife-cloud-1.2.0/lib/chef/knife/cloud/command.rb:51:in `run'
        from /opt/chefdk/embedded/apps/chef/lib/chef/knife.rb:425:in `block in run_with_pretty_exceptions'
        from /opt/chefdk/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
        from /opt/chefdk/embedded/apps/chef/lib/chef/knife.rb:424:in `run_with_pretty_exceptions'
        from /opt/chefdk/embedded/apps/chef/lib/chef/knife.rb:215:in `run'
        from /opt/chefdk/embedded/apps/chef/lib/chef/application/knife.rb:142:in `run'
        from /opt/chefdk/embedded/apps/chef/bin/knife:25:in `<top (required)>'
        from /usr/bin/knife:55:in `load'
        from /usr/bin/knife:55:in `<main>'

( ccorbett@cherenkov:~/development/cookbooks/sce-vcloud (feature) )$ knife vcair server create --help | grep vcair-net
        --vcair-net NETWORKNAME      Your VCAIR NETWORK

( ccorbett@cherenkov:~/development/cookbooks/sce-vcloud (feature) )$ gem list | grep vcair
knife-vcair (0.7.0)
```

**Comment by adamleff** [comment]
@asciifaceman thanks for the issue report.  The error handling here could definitely be better.  In this case, it appears that the plugin is trying to find the network you specified in your config / CLI options but cannot find it, and rather than tell you as such, we throw a useless exception.  I'll post up a PR to raise a better exception in this case.

However, that won't solve your issue.  Is it possible that you're specifying the name of the network incorrectly?  It is case-sensitive, and unfortunately, the vCloud Air UI is not helpful in that it generally shows everything as upper-case.

Here's a quick ruby script you can use to list out the networks available to you in your vcair org in hopes it might shed a light on the network name issue:

```ruby
require 'fog'

username = 'user@domain.com'
password = 'soopers33kret'
org      = 'some-uuid-here'
api_host = 'api-endpoint.vmware.com'

auth_params = {
  provider: 'vclouddirector',
  vcloud_director_username: "#{username}@#{org}",
  vcloud_director_password: password,
  vcloud_director_host: api_host,
  path: '/api/compute/api'
}

service = Fog::Compute.new(auth_params)
org     = service.organizations.get_by_name(org)
puts org.networks.map(&:name)
```

Let me know if that helps or if you're still running into an issue.

**Comment by cstewart87** [comment]
@asciifaceman I believe I've seen something like this in the past if you have multiple networks with the same name (all VDC default networks have the same name).

The code base (https://github.com/chef-partners/knife-vcair/blob/master/lib/chef/knife/vcair_helpers.rb#L30) searches for networks by name.  I don't recall if fog currently supports a `get_by_id` for vca networks, but finding by ID could help (if this turns out to be the issue).

What I would recommend is to first verify if you have multiple networks with the same name (per @adamleff's recommendation); if so try editing the name of the network you're using to be unique across your account.

**Comment by adamleff** [comment]
Great to see you chiming in here, @cstewart87 - thanks!

Fog does, indeed, support searching networks by ID.  I use it in kitchen-vcair: https://github.com/chef-partners/kitchen-vcair/blob/master/lib/kitchen/driver/vcair.rb#L217-L223

If your hypothesis is correct, I will put up a PR with a better exception message as well as support for specifying the network by ID or name.

**Comment by asciifaceman** [comment]
I only have one network defined at the moment, "poc-network" inside a VDC named "sce-poc".  Based on other conversations I wonder if I am even hitting the right endpoint, which assembled should look like us-california-1-3.vchs.vmware.com/api/compute/api

**Comment by asciifaceman** [comment]
@adamleff your scan script actually worked for me... But even with knife vcair, I could view/list networks and hosts. I just can't seem to create anything.

**Comment by adamleff** [comment]
@asciifaceman So, the script showed your network, and using that script, you've confirmed that you've set your knife parameters the same way, specifying the network name with the correct case/capitalization, and still getting the NilClass exception?

**Comment by asciifaceman** [comment]
yes, I can run knife vcair network list, server list, vm list. However I cannot run server create without the NilClass exception.

**Comment by adamleff** [comment]
Here's an updated test script:

```ruby
require 'fog'

username = 'user@domain.com'
password = 'soopers33kret'
org      = 'some-uuid-here'
api_host = 'api-endpoint.vmware.com'
network_name = '<NETWORK NAME HERE>'

auth_params = {
  provider: 'vclouddirector',
  vcloud_director_username: "#{username}@#{org}",
  vcloud_director_password: password,
  vcloud_director_host: api_host,
  path: '/api/compute/api'
}

service = Fog::Compute.new(auth_params)
org     = service.organizations.get_by_name(org)
puts '### Available Networks'
puts org.networks.map(&:name)
puts ''
puts '### Found Network'
puts org.networks.get_by_name(network_name).inspect

```

Can you provide the output of that test script I provided to you above, the vcair-related items in your knife.rb (with your password redacted), and the exact `knife vcair server create` command you are trying to execute?

If there's nothing obvious there, I think the next step is for me to release a new gem with some better exceptions showing what network name knife-vcair is seeing and what networks are actually available in hopes we see where the mismatch is.  Right now, the only thing that's clear is that knife-vcair is seeing that you wish to use a network, but Fog is not able to find that network.

**Comment by asciifaceman** [comment]
```
require 'fog'

username = 'redacted'
password = 'redacted'
org      = 'redacted'
api_host = 'us-california-1-3.vchs.vmware.com'
network_name = 'tenancy01-poc'

auth_params = {
  provider: 'vclouddirector',
  vcloud_director_username: "#{username}@#{org}",
  vcloud_director_password: password,
  vcloud_director_host: api_host,
  path: '/api/compute/api'
}

service = Fog::Compute.new(auth_params)
org     = service.organizations.get_by_name(org)
puts '### Available Networks'
puts org.networks.map(&:name)
puts ''
puts '### Found Network'
puts org.networks.get_by_name(network_name).inspect
```

```
### Available Networks
tenancy01-poc
poc-network
default-routed-network

### Found Network
  <Fog::Compute::VcloudDirector::Network
    id="redacted",
    name="tenancy01-poc",
    type="application/vnd.vmware.vcloud.orgVdcNetwork+xml",
    href="https://us-california-1-3.vchs.vmware.com/api/compute/api/admin/network/redacted",
    description="Attempt to POC existing tenancy structure",
    is_inherited="false",
    is_shared="false",
    fence_mode="natRouted",
    gateway="10.0.1.1",
    netmask="255.255.255.0",
    dns1="8.8.8.8",
    dns2="8.8.4.4",
    dns_suffix=nil,
    ip_ranges=[{:start_address=>"10.0.1.201", :end_address=>"10.0.1.254"}]
  >

```

Odd, it seems to have no problem finding the network

**Comment by asciifaceman** [comment]
Additionally in my knife.rb the api_host, user, pass, org, and api_path are all defined (identical to this script)

**Comment by adamleff** [comment]
@asciifaceman I have pushed knife-vcair-0.8.0 up to rubygems.  Please upgrade at your convenience and attempt your `knife vcair server create` again.  Please ensure you are running with `-VV` to pick up the additional debug logging I've added.  Thanks!

**Comment by asciifaceman** [comment]
With 0.8.0:

```
DEBUG: Using fog version: 1.34.0
DEBUG: Looking up network by name: tenancy01-poc
DEBUG: Using network tenancy01-poc (6bf64897-1577-4306-9fe9-17228c1fa24f)
/home/ccorbett/.chefdk/gem/ruby/2.1.0/gems/knife-vcair-0.8.0/lib/chef/knife/vcair_server_create.rb:145:in `instantiate': undefined method `instantiate' for nil:NilClass (NoMethodError)
        from /home/ccorbett/.chefdk/gem/ruby/2.1.0/gems/knife-vcair-0.8.0/lib/chef/knife/vcair_server_create.rb:36:in `execute_command'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/knife-cloud-1.2.0/lib/chef/knife/cloud/command.rb:51:in `run'
        from /opt/chefdk/embedded/apps/chef/lib/chef/knife.rb:425:in `block in run_with_pretty_exceptions'
        from /opt/chefdk/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
        from /opt/chefdk/embedded/apps/chef/lib/chef/knife.rb:424:in `run_with_pretty_exceptions'
        from /opt/chefdk/embedded/apps/chef/lib/chef/knife.rb:215:in `run'
        from /opt/chefdk/embedded/apps/chef/lib/chef/application/knife.rb:142:in `run'
        from /opt/chefdk/embedded/apps/chef/bin/knife:25:in `<top (required)>'
        from /usr/bin/knife:55:in `load'
        from /usr/bin/knife:55:in `<main>'
```

**Comment by asciifaceman** [comment]
Seems to be failing on the template.instantiate line of:

```
        def instantiate
          begin
            node_name = config_value(:chef_node_name)
            template.instantiate(
                                 node_name,
                                 vdc_id: vdc.id,
                                 network_id: net.id,
                                 description: "id:#{node_name}")
          rescue CloudExceptions::ServerCreateError => e
            raise e
          end
        end
```

**Comment by adamleff** [comment]
So, this is new and different from what you initially reported. But we appear to be past the network issue.  :) 

next week I'll add similar exception handling for the template method and that should help. Talk to you then.

**Comment by asciifaceman** [comment]
thank you :)

I guess I didn't have the depth of field to tell what the actual problem is. Do you want me to close / recreate the issue or will you return to this?

**Comment by adamleff** [comment]
Let's keep working off this issue until you're fixed. No need for another one.

**Comment by asciifaceman** [comment]
kk, see you next week.

**Comment by adamleff** [comment]
@asciifaceman I have submitted PR #35 which should give us more information about the template issue you're experiencing.  If you're familiar with how to use bundler and point it to a specific repo/branch, please feel free to point at my branch and test.  Otherwise, once this PR is merged (hopefully later today), I'll cut a new version of knife-vcair and publish it to rubygems.

**Comment by asciifaceman** [comment]
I was going to pull it using git in my Gemfile but it was throwing some errors I am not used to seeing so I can just wait until the gem update xD

**Comment by asciifaceman** [comment]
Looks like VMWare broke vDirector over the weekend too (SSO auth from vmware is broken it looks like) so I can't do any network config anyways

**Comment by adamleff** [comment]
@asciifaceman I have released 0.9.0 of the knife-vcair gem with better template exception handling.  Please try it out and let me know what you see now.

**Comment by asciifaceman** [comment]
oddly now it is actually working almost, it's spinning up vm's, but also giving the error:
ERROR: Fog::Compute::VcloudDirector::TaskError: status: error, error: {:majorErrorCode=>"400", :message=>"[ 63a9df37-e52f-4394-8a5f-5958addefbf9 ] The VCD entity network \"stg\" does not exist.", :minorErrorCode=>"BAD_REQUEST"}

And exiting out. I am not defining "stg" anywhere...

**Comment by adamleff** [comment]
Nothing in knife-vcair has that string either, so that's something specific to your environment and Fog's interaction with vCloud Air.  I can try to debug it, but I'll need you to run with `-VV` so I can see a stacktrace as to where that's coming from.

**Comment by asciifaceman** [comment]
```
( ccorbett@cherenkov:~/development )$ knife vcair server create --ssh-password ***** --image "debian_jessie_base" --node-name "******" --vcair-net "*****(not stg)" -VV
WARN: Unresolved specs during Gem::Specification.reset:
      fog (~> 1.33)
      little-plugger (>= 1.1.3)
      mini_portile (~> 0.6.0) # I am wondering if this might the issue, these three specs
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
INFO: Using configuration from /home/ccorbett/.chef/knife.rb
INFO: Started chef-zero at chefzero://localhost:8889 with repository at /home/ccorbett/development
  One version per cookbook

DEBUG: vcair_username oops
DEBUG: vcair_org oops
DEBUG: vcair_api_host us-california-1-3.vchs.vmware.com
DEBUG: vcair_api_path /api/compute/api
DEBUG: vcair_api_version 
DEBUG: vcair_show_progress 
DEBUG: Using fog version: 1.34.0
DEBUG: Searching catalogs for image debian_jessie_base...
DEBUG: Searching catalog Public Catalog...
DEBUG: Image not found in catalog Public Catalog - possible images: CentOS64-64BIT, CentOS64-32BIT, Ubuntu Server 12.04 LTS (amd64 20150127), Ubuntu Server 12.04 LTS (i386 20150127), W2K12-STD-64BIT, CentOS63-64BIT, VMware Photon OS - Tech Preview 2, W2K12-STD-R2-64BIT, W2K8-STD-R2-64BIT, CentOS63-32BIT
DEBUG: Searching catalog default-catalog...
DEBUG: Image debian_jessie_base (739effee-6f58-4eee-a235-9ed4dc8d715a) found in catalog default-catalog - search complete.
DEBUG: Looking up network by name: tenancy01-poc
DEBUG: Using network tenancy01-poc (6bf64897-1577-4306-9fe9-17228c1fa24f)
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/fog-1.34.0/lib/fog/vcloud_director/compute.rb:436:in `wait_and_raise_unless_success': status: error, error: {:majorErrorCode=>"400", :message=>"[ f190d1b2-30c6-414b-90fa-73532a5eccdc ] The VCD entity network \"stg\" does not exist.", :minorErrorCode=>"BAD_REQUEST"} (Fog::Compute::VcloudDirector::TaskError)
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/fog-1.34.0/lib/fog/vcloud_director/compute.rb:425:in `process_task'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/fog-1.34.0/lib/fog/vcloud_director/models/compute/catalog_item.rb:17:in `instantiate'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/knife-vcair-0.9.0/lib/chef/knife/vcair_server_create.rb:145:in `instantiate'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/knife-vcair-0.9.0/lib/chef/knife/vcair_server_create.rb:36:in `execute_command'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/knife-cloud-1.2.0/lib/chef/knife/cloud/command.rb:51:in `run'
        from /opt/chefdk/embedded/apps/chef/lib/chef/knife.rb:425:in `block in run_with_pretty_exceptions'
        from /opt/chefdk/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
        from /opt/chefdk/embedded/apps/chef/lib/chef/knife.rb:424:in `run_with_pretty_exceptions'
        from /opt/chefdk/embedded/apps/chef/lib/chef/knife.rb:215:in `run'
        from /opt/chefdk/embedded/apps/chef/lib/chef/application/knife.rb:142:in `run'
        from /opt/chefdk/embedded/apps/chef/bin/knife:25:in `<top (required)>'
        from /usr/bin/knife:55:in `load'
        from /usr/bin/knife:55:in `<main>'

```

**Comment by adamleff** [comment]
@asciifaceman I'm not seeing anything in the knife-vcair code that would cause that.  The error is coming from Fog so it's possible that Fog has an issue dealing with custom VCDs.  I'm not entirely sure as I'm not able to reproduce this.

I'm going to close out this issue since I believe we have addressed the main problems you reported and added some additional debug output to help future users.  If we can help again, just let us know.

**Comment by cstewart87** [comment]
@asciifaceman I know this issue is closed but it might be worth opening another issue against the fog repo.  

Have you tried specifying your VCD using `--vcair-vdc` just to ensure you're using the correct VCD?  (Although, if you already have it set in your `knife.rb` then disregard my comment :smile: )

Also, you might want to check out our [vmworld-chef-repo](https://github.com/vmwaredevops/vmworld-chef-repo) repository for additional examples.  If you have any questions or issues from there feel free to submit an issue directly on that repo.

Hope at least some of this helps!

**Comment by adamleff** [comment]
@cstewart87 I love that you're participating in this repo and when issues come up!  Thanks for helping!

**Comment by cstewart87** [comment]
@adamleff Not a problem!  I've actually been having some recent thoughts about breaking out a fog-vcloud_air library for specific vcloud_air support.  It would be a large effort but might help with vcair integration for some tools like knife-vcair.

**Comment by adamleff** [comment]
@cstewart87 Fog is actually pretty awesome but has grown in size/complexity to the point where some of our recent projects have involved us writing a standalone ruby client gem specifically for that need.  It might be time for a vclouddirector gem that is solely focused on that functionality.

**Comment by asciifaceman** [comment]
If it helps, this project was make or break for my department at Sony to
adopt vCloud. So far It was break, existing tools are just too immature for
us to pursue a code based deployment infrastructure on vCloud when 900%
better options exist. :(

**Comment by cstewart87** [comment]
@asciifaceman Sorry to hear that!  I imagine timing is a factor here, but I know these tools will continue to improve.  I'll reach out to some of my peers to determine how we can help improve integrations with vCloud Air

**Comment by mreferre** [comment]
I have skammed quickly through this issue so I am not sure if I am adding confusion or something useful. Anyway.... here is how I was able to deploy VMs from templates and connect them to Org Networks inside vCloud Air using Fog: http://it20.info/2015/09/fog-io-and-vcloud-air/ 

Note the networking caveats at the end (probably not applicable to this issue). Note also the apipath thingy. 

HTH
